server/datastore/mysql/operating_systems.go

```diff
@@ -0,0 +1,144 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jmoiron/sqlx"
+)
+
+func (ds *Datastore) ListOperatingSystems(ctx context.Context) ([]fleet.OperatingSystem, error) {
+	return listOperatingSystemsDB(ctx, ds.reader)
+}
+
+func listOperatingSystemsDB(ctx context.Context, tx sqlx.QueryerContext) ([]fleet.OperatingSystem, error) {
+	var os []fleet.OperatingSystem
+	if err := sqlx.SelectContext(ctx, tx, &os, `SELECT * FROM operating_systems`); err != nil {
+		return nil, err
+	}
+	return os, nil
+}
+
+func (ds *Datastore) UpdateHostOperatingSystem(ctx context.Context, hostID uint, hostOS fleet.OperatingSystem) error {
+	os, err := maybeNewOperatingSystemDB(ctx, ds.writer, hostOS)
+	if err != nil {
+		return err
+	}
+	return maybeUpdateHostOperatingSystemDB(ctx, ds.writer, hostID, os.ID)
+}
+
+// maybeNewOperatingSystemDB queries the `operating_systems` table with the
+// name, version, arch, and kernel_version of the given operating system. If found,
+// it returns the record including the associated ID. If not found, it returns a call
+// to `newOperatingSystemDB`, which inserts a new record and returns the record
+// including the newly associated ID.
+func maybeNewOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	switch os, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return os, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return newOperatingSystemDB(ctx, tx, hostOS)
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get operating system")
+	}
+}
+
+// `newOperatingSystemDB` inserts a record for the given operating system and
+// returns the record including the newly associated ID.
+func newOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	stmt := "INSERT INTO operating_systems (name, version, arch, kernel_version) VALUES (?, ?, ?, ?)"
+	if _, err := tx.ExecContext(ctx, stmt, hostOS.Name, hostOS.Version, hostOS.Arch, hostOS.KernelVersion); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "insert new operating system")
+	}
+
+	switch storedOS, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return storedOS, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return nil, doRetryErr
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get new operating system")
+	}
```

Comment:
Do we need to query the db again to get the ID? We should be able to use `Result.LastInsertId()` .

Reply:
I wondered that as well. This approach was based on the following comment from `getOrGenerateSoftwareIdDB`:
```
// LastInsertId sometimes returns 0 as it's dependent on connections and how mysql is
// configured.
```

Reply:
This comment seems a bit misleading. I think the author, was running into an issue with `INSERT ... ON DUPLICATE KEY UPDATE` returning a 0 when an update occurred. A non-zero last insert id would only be returned if there was an insert. @lucasmrod would you mind clarifying?

Reply:
Yes, the 0 return was because of an update instead of an insert. It should be safe to do a `LastInsertId` here. If we change the insert to have any `ON DUPLICATE ...` then we would need to check for 0 in that case.

Reply:
Correct, it's due to the `INSERT` **with** `ON DUPLICATE KEY UPDATE`.